# Regression

In [1]:
import pandas as pd
import numpy as np
import bq_helper
from bq_helper import BigQueryHelper
# https://www.kaggle.com/sohier/introduction-to-the-bq-helper-package
genomes = bq_helper.BigQueryHelper(active_project="bigquery-public-data",
                                   dataset_name="genomics_cannabis")

In [2]:
bq_assistant = BigQueryHelper("bigquery-public-data", "genomics_cannabis")
bq_assistant.list_tables()

['MNPR01_201703',
 'MNPR01_reference_201703',
 'MNPR01_transcriptome_201703',
 'cs10_gff',
 'cs3k_project_info',
 'cs3k_vcf_cs10_dv090',
 'sample_info_201703']

In [3]:
query = """SELECT
  variant_id, alternate_bases, quality, type, 
  AB, ABP, AC, AF, AN, AO, DP, DPB, DPRA, EPP, 
  EPPR, GTI, MQM, MQMR, NS, NUMALT, ODDS, PAO,
  PQA, PQR, PRO, QA, QR, RO, RPL, RPP, RPPR, 
  RPR, RUN, SAF, SAP, SAR, SRF, SRP, SRR
FROM
  `bigquery-public-data.genomics_cannabis.MNPR01_201703` v
limit 1000000;"""
response = genomes.query_to_pandas_safe(query, max_gb_scanned=100)
response.head(10)

C:\Users\sharm\AppData\Roaming\Python\Python311\site-packages\google\cloud\bigquery\table.py:1727: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.
  warnings.warn(


variant_id alternate_bases  \
0  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTI1NzF8Z2J8TU5QUj...           [CTA]   
1  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTEyODV8Z2J8TU5QUj...         [CGTAA]   
2  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTY5MTJ8Z2J8TU5QUj...          [CCTT]   
3  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTY5OTl8Z2J8TU5QUj...         [TATGG]   
4  CKXG8eKP9qOf8wESIGdpfDEwOTg0ODY2OTF8Z2J8TU5QUj...    [ATTTTTTTAA]   
5  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTQyOTF8Z2J8TU5QUj...           [GAG]   
6  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTUwNTV8Z2J8TU5QUj...            [GC]   
7  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTUyMDF8Z2J8TU5QUj...        [GATAAA]   
8  CKXG8eKP9qOf8wESIGdpfDEwOTg0OTU1NTV8Z2J8TU5QUj...        [TTTTGC]   
9  CKXG8eKP9qOf8wESIGdpfDEwOTg0ODgxOTZ8Z2J8TU5QUj...           [AGT]   

    quality       type     AB    ABP   AC     AF  AN   AO  ...        RPP  \
0  23.39620      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
1   7.52241  [complex]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
2  25.30020      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
3   6.16805  [complex]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
4   3.08123  [complex]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
5  23.56570      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
6  27.00690      [del]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
7  14.01670      [mnp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
8  12.62600      [mnp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   
9   9.12466      [snp]  [0.0]  [0.0]  [2]  [1.0]   2  [1]  ...  [5.18177]   

   RPPR  RPR  RUN  SAF        SAP  SAR  SRF  SRP  SRR  
0   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
1   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
2   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
3   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
4   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
5   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
6   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
7   0.0  [0]  [1]  [0]  [5.18177]  [1]    0  0.0    0  
8   0.0  [0]  [1]  [1]  [5.18177]  [0]    0  0.0    0  
9   0.0  [1]  [1]  [0]  [5.18177]  [1]    0  0.0    0  

[10 rows x 39 columns]

I shall run regression to determine my DNA samples based on proteins. 

In [4]:
response["DNA_A"] = response["alternate_bases"].str[0].str.count('A')

In [5]:
response["DNA_C"] = response["alternate_bases"].str[0].str.count('C')

In [6]:
response["DNA_G"] = response["alternate_bases"].str[0].str.count('G')

In [7]:
response["DNA_T"] = response["alternate_bases"].str[0].str.count('T')

Now I shall adjust the proteins for easy analysis. 

In [8]:
proteins = ["AB", "ABP", "AC", "AF", "AO", "DPRA", "EPP", "MQM", "PAO",
            "PQA", "QA", "RPL", "RPP", 
            "RPR", "RUN", "SAF", "SAP", "SAR", 
]

for protein in proteins:
    print(protein)
    response[protein] = response[protein].str[0]

AB
ABP
AC
AF
AO
DPRA
EPP
MQM
PAO
PQA
QA
RPL
RPP
RPR
RUN
SAF
SAP
SAR


In [9]:
response["type"] = response["type"].str[0]

In [10]:
response["type"].value_counts()

type
snp        699662
complex    158031
mnp         57898
ins         51333
del         33076
Name: count, dtype: int64

```type``` is a categorical variable, so I shall need to dummify it. 

In [11]:
dummies = pd.get_dummies(response["type"], dtype="int")
dummies

complex  del  ins  mnp  snp
0             0    0    0    0    1
1             1    0    0    0    0
2             0    0    0    0    1
3             1    0    0    0    0
4             1    0    0    0    0
...         ...  ...  ...  ...  ...
999995        0    0    0    0    1
999996        1    0    0    0    0
999997        0    0    0    0    1
999998        1    0    0    0    0
999999        0    0    0    0    1

[1000000 rows x 5 columns]

Now I shall define what I am regressing upon.

In [12]:
y = response[["DNA_A", "DNA_C", "DNA_G", "DNA_T"]]

In [13]:
X = pd.concat([
    response[["AB", "ABP", "AC", "AF", "AN", "AO", "DP", "DPB", "DPRA", "EPP", 
              "EPPR", "GTI", "MQM", "MQMR", "NS", "NUMALT", "ODDS", "PAO",
              "PQA", "PQR", "PRO", "QA", "QR", "RO", "RPL", "RPP", "RPPR", 
              "RPR", "RUN", "SAF", "SAP", "SAR", "SRF", "SRP", "SRR"
    ]], 
    dummies
], axis=1)

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.33, random_state=1337)
LR = LinearRegression()
LR.fit(X_train, y_train)
train_score = LR.score(X_train, y_train)
test_score = LR.score(X_test, y_test)
print(f'Train score {train_score}, test score {test_score}')

Train score 0.17432333633747393, test score 0.15232295290835307


I've achieved a test score of 15.2%. This is unsatisfactory, I will need a different method. 